[WebGPU] Vertex buffer support

mwyrzykowski opened the pull request:
#### d2a1f86a1fae6997f383e24f73afbf080f142d53
<pre>
[WebGPU] Vertex buffer support
<a href="https://bugs.webkit.org/show_bug.cgi?id=248071">https://bugs.webkit.org/show_bug.cgi?id=248071</a>
&lt;radar://101564178&gt;

Reviewed by Dean Jackson.

Implement basic vertex buffer support which supports the rotating cube test case.

Very preliminary bind group support is implemented, which only
supports 1 buffer of arbitrary size.

* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
(WebCore::GPUBuffer::GPUBuffer):
(WebCore::GPUBuffer::getMappedRange):
(WebCore::GPUBuffer::unmap):
If the buffer is mappedAtCreation then the API allows for
writing to the mapped range even if the write flag is not set.

We need to store the JSC::ArrayBuffer too so we
can copy it back to the MTLBuffer when the client
calls GPUBuffer.unmap

* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
Preliminary support for bind groups via tier2 argument
buffers which are substantially easier to implement as
the tier1 path was not functioning in my testing.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setVertexBuffer):
(WebGPU::RenderPassEncoder::setViewport):
Implement setting vertex buffer and bind groups.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::getMappedRange):
(WebKit::RemoteBuffer::unmap):
If the buffer is mappedAtCreation then the API  allows for
writing to the mapped range even if the write flag is not set.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.messages.in:
* Source/WebKit/Shared/WebGPU/WebGPUBindGroupEntry.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
Need to pass along a valid identifier unless the auto-generated
serializer asserts.

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::getMappedRange):
Support getMappedRange API.

* Websites/webkit.org/demos/webgpu/hello-cube.html:
* Websites/webkit.org/demos/webgpu/scripts/hello-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Use requestAnimationFrame to spin the cube.

Canonical link: <a href="https://commits.webkit.org/257187@main">https://commits.webkit.org/257187@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e02b326c389742d906e98eebdd5808bae738b3fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98090 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31239 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107551 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167815 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102030 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7793 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36067 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90648 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104159 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103742 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5861 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84686 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32961 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87730 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89445 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88941 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1277 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84663 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1240 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22380 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28618 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4948 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6102 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87501 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2542 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41786 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19627 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
```diff
@@ -47,12 +53,15 @@ void GPUBuffer::mapAsync(GPUMapModeFlags mode, std::optional<GPUSize64> offset,
 {
-    auto mappedRange = m_backing->getMappedRange(offset.value_or(0), size);
-    return ArrayBuffer::create(mappedRange.source, mappedRange.byteLength);
+    m_mappedRange = m_backing->getMappedRange(offset.value_or(0), size);
```
Do we need to validate that size is <= the size of the buffer?

mwyrzykowski replied:
This is validated in WebGPU.framework layer as getMappedRange could be called by the native client.

Do you think we should validate in the web process layer too? It would be double validation and does not seem to occur for the other calls in WebCore from what I can tell

grorg replied:
Maybe put a comment in about this, in case our Security folks question it when they do a code inspection?

grorg replied:
Hmm... I suck at github. That comment is marked as Pending, but it is just a suggestion.

mwyrzykowski replied:
I added a comment :+1:

djg commented on Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
```diff
@@ -47,12 +53,15 @@ void GPUBuffer::mapAsync(GPUMapModeFlags mode, std::optional<GPUSize64> offset,
 
 void GPUBuffer::unmap()
 {
+    memcpy(m_mappedRange.source, m_arrayBuffer->data(), m_mappedRange.byteLength);
```
Should the JS array buffer be reset after the copy?

mwyrzykowski replied:
Yes I will address that

djg commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -55,6 +55,32 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
+            bool textureViewIsPresent = WebGPU::textureViewIsPresent(entry);
+            if (bufferIsPresent) {
+                id<MTLBuffer> buffer = WebGPU::fromAPI(entry.buffer).buffer();
+                *(float**)vertexArgumentBuffer.contents = (float*)buffer.gpuAddress;
```
`gpuAddress` is available from macOS 13 and iOS 16. Do we need to add version guards around its use or should WebGPU not be compiled for platforms before these?

mwyrzykowski replied:
I will add some form of version guard for now and eventually we may likely want to support tier1 argument buffer devices where `gpuAddress` is not needed

grorg replied:
I guess this is why the build failed?

grorg replied:
Yeah.

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
```diff
@@ -28,6 +28,12 @@
 
 namespace WebCore {
 
+GPUBuffer::GPUBuffer(Ref<PAL::WebGPU::Buffer>&& backing)
```
this change is not needed, I will revert it

mwyrzykowski committed:
[WebGPU] Vertex buffer support
https://bugs.webkit.org/show_bug.cgi?id=248071
<radar://101564178>

Reviewed by Dean Jackson.

Implement basic vertex buffer support which supports the rotating cube test case.

Very preliminary bind group support is implemented, which only
supports 1 buffer of arbitrary size.

* Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
(WebCore::GPUBuffer::GPUBuffer):
(WebCore::GPUBuffer::getMappedRange):
(WebCore::GPUBuffer::unmap):
If the buffer is mappedAtCreation then the API allows for
writing to the mapped range even if the write flag is not set.

We need to store the JSC::ArrayBuffer too so we
can copy it back to the MTLBuffer when the client
calls GPUBuffer.unmap

* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
Preliminary support for bind groups via tier2 argument
buffers which are substantially easier to implement as
the tier1 path was not functioning in my testing.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setVertexBuffer):
(WebGPU::RenderPassEncoder::setViewport):
Implement setting vertex buffer and bind groups.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::getMappedRange):
(WebKit::RemoteBuffer::unmap):
If the buffer is mappedAtCreation then the API  allows for
writing to the mapped range even if the write flag is not set.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.messages.in:
* Source/WebKit/Shared/WebGPU/WebGPUBindGroupEntry.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
Need to pass along a valid identifier unless the auto-generated
serializer asserts.

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::getMappedRange):
Support getMappedRange API.

* Websites/webkit.org/demos/webgpu/hello-cube.html:
* Websites/webkit.org/demos/webgpu/scripts/hello-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Use requestAnimationFrame to spin the cube.

Canonical link: https://commits.webkit.org/257187@main
